include/WaveParallel.hpp

```diff
@@ -0,0 +1,338 @@
+#ifndef WAVE_EQUATION_PARALLEL_HPP
+#define WAVE_EQUATION_PARALLEL_HPP
+
+
+// ==================== INCLUDES ====================
+
+// Logging in parallel environment
+#include <deal.II/base/conditional_ostream.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/base/function.h>
+#include <deal.II/base/utilities.h>
+
+// Distributed Triangulation and Grid Generation
+#include <deal.II/distributed/fully_distributed_tria.h>
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_in.h>
+
+// Finite Element space
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/fe_simplex_p.h>
+#include <deal.II/fe/fe_q.h>
+
+// Degrees of freedom
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+
+// Linear Algebra - Distributed
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/trilinos_precondition.h>
+#include <deal.II/lac/trilinos_sparse_matrix.h>
+
+// Numeric techniques
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/numerics/matrix_tools.h>
+
+// Timing
+#include <deal.II/base/timer.h>
+
+// STL libraries
+#include <iostream>
+#include <fstream>
+#include <memory>
+
+// ==================================================
+
+using namespace dealii;
+
+template <unsigned int dim>
+class WaveEquationParallel
+{
+public:
+    /**
+     * @brief Constructs the triangulation, finite element space, DoF handler and linear algebra,
+     * utilising a mesh file passed as input.
+     * @param mesh_filename Path to the mesh file.
+     */
+    void setup(const std::string& mesh_filename);
+
+    /**
+     * @brief Constructs the triangulation, finite element space, DoF handler and linear algebra,
+     * utilising deal.ii mesh generation infrastructure.
+     * 
+     * @param times Number of times the mesh will be refined.
+     */
+    void setup(const unsigned int& times=1);
```

Comment:
As it turns out, these two methods do quite different things: the first reads a tetrahedral mesh from file, whereas the second builds a square mesh of hexahedra.

In this situation, I would refrain from relying on overloading. The most typical use case of overloading is when you have a function or method that does the same thing on different inputs. Here, you do different things with different inputs: it would probably be better to use different function names as well.